# INTRODUCCIÓN A DEEP LEARNING
## Trabajo adaptado por Lizeth Solano Romo, al trabajo original del Profesor Benjamin Oliva Vazquez, 1 mayo 2024

## ¿Qué es el aprendizaje profundo?

* Algunos de los avances más impresionantes en inteligencia artificial de los últimos años se han producido en el campo del aprendizaje profundo.

* El procesamiento del lenguaje natural, el reconocimiento de imágenes y los juegos son tareas en las que los modelos de aprendizaje profundo se han acercado o incluso superado el rendimiento a nivel humano.

* ¿Qué es el aprendizaje profundo? El aprendizaje profundo es un enfoque del aprendizaje automático que ha permitido identificar tipos de patrones complejos y jerárquicos que se encuentran en los conjuntos de datos más desafiantes del mundo real.

El objetivo principal es presentar un proyecto sencillo de Deep Learning y practicar los principales usos de bibliotecas como `keras` o `tensorflow`.

# Load Data


In [ ]:
#!pip install tensorflow

In [8]:
# first neural network with keras tutorial
from numpy import loadtxt
from numpy import genfromtxt
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

#
import warnings
warnings.filterwarnings('ignore')

En esta clase de Keras, utilizaremos el conjunto de datos sobre la percepción de las mujeres con relación a las actitudes y justificaciones para cometer actos de violencia contra ellas.

Este es un conjunto de datos de aprendizaje automático estándar del repositorio [Kaggle](https://www.kaggle.com/datasets/whenamancodes/violence-against-women-girls/data).

Los datos que contiene esta BD, se tomaron de una encuesta realizada a hombres y mujeres en países africanos, asiáticos y sudamericanos, explorando las actitudes y las justificaciones percibidas dadas para cometer actos de violencia contra las mujeres. Los datos también exploran diferentes grupos sociodemográficos a los que pertenecen los encuestados, que incluyen: nivel de educación, estado civil, empleo y grupo de edad.

Sin embargo, para este ejercicio solo se tomaron en considereación las observaciones realizadas por mujeres (desde una perspectiva feminista), y en particular, para países de SudAmérica. Esto con la intención de evidenciar la gravedad de la situación en el continente Americano y establecer campañas de concientización pública y al acceso de las mujeres y las niñas a la educación para comunicar que la violencia contra ellas, nunca es aceptable ni justificable.

Para realizar este ejercicio, se realizaron ajustes a la BD para que todas las variables de entrada que describen a cada paparticipante fueran numéricas, facilitando su uso directamente con redes neuronales que esperan valores numéricos de entrada y salida y es una opción ideal para nuestra primera red neuronal en Keras.

El conjunto de datos está disponible [aquí](hhttps://www.kaggle.com/datasets/whenamancodes/violence-against-women-girls/data). Es un archivo CSV del conjunto de datos (VAGG.csv). Los detalles del conjunto de datos esta disponible [aquí](https://raw.githubusercontent.com/lisolano/ModuloVI/main/VAGG.csv).

Las variables usadas en la BD son:

Variables de entrada (X):

* Country	= país
* Demographics Question	= pregunta demográfica
* Demographics Response	= respuesta demográfica
* Question-cod	=código de la pregunta
* year	= año
* Value = valor

Variable de salida (y):
* SA (0 or 1) donde 1 es SudAmérica y 0 países en otros continentes


In [9]:
#

dataset = loadtxt('https://raw.githubusercontent.com/lisolano/ModuloVI/main/VAGG.csv', delimiter=',', skiprows=1)

dataset

array([[1.000e+00, 4.000e+00, 8.000e+00, ..., 2.015e+03, 0.000e+00,
        0.000e+00],
       [1.000e+00, 2.000e+00, 6.000e+00, ..., 2.015e+03, 4.500e+00,
        0.000e+00],
       [1.000e+00, 5.000e+00, 1.400e+01, ..., 2.015e+03, 4.600e+00,
        0.000e+00],
       ...,
       [7.000e+01, 5.000e+00, 1.400e+01, ..., 2.015e+03, 1.500e+01,
        0.000e+00],
       [7.000e+01, 5.000e+00, 1.100e+01, ..., 2.015e+03, 7.200e+00,
        0.000e+00],
       [7.000e+01, 5.000e+00, 1.400e+01, ..., 2.015e+03, 3.700e+00,
        0.000e+00]])

In [10]:
# split into input (X) and output (y) variables
X = dataset[:,0:6]
y = dataset[:,6]

In [11]:
X

array([[1.000e+00, 4.000e+00, 8.000e+00, 1.000e+00, 2.015e+03, 0.000e+00],
       [1.000e+00, 2.000e+00, 6.000e+00, 1.000e+00, 2.015e+03, 4.500e+00],
       [1.000e+00, 5.000e+00, 1.400e+01, 1.000e+00, 2.015e+03, 4.600e+00],
       ...,
       [7.000e+01, 5.000e+00, 1.400e+01, 4.000e+00, 2.015e+03, 1.500e+01],
       [7.000e+01, 5.000e+00, 1.100e+01, 6.000e+00, 2.015e+03, 7.200e+00],
       [7.000e+01, 5.000e+00, 1.400e+01, 6.000e+00, 2.015e+03, 3.700e+00]])

In [12]:
y

array([0., 0., 0., ..., 0., 0., 0.])

## Define Keras Model

Los modelos en Keras se definen como una secuencia de capas.

Creamos un modelo secuencial y agregamos capas una a la vez hasta que estemos satisfechos con nuestra arquitectura de red.

Lo primero que hay que hacer es asegurarse de que la capa de entrada tenga la cantidad correcta de entidades de entrada. Esto se puede especificar al crear la primera capa con el argumento input_shape y configurarlo en (5,) para presentar las cinco variables de entrada como un vector.

### ¿Cómo sabemos el número de capas y sus tipos?

Esta es una pregunta complicada.

Hay heurísticas que puede utilizar y, a menudo, la mejor estructura de red se encuentra mediante un proceso de experimentación de prueba y error.

Generalmente, necesita una red lo suficientemente grande como para capturar la estructura del problema.

En este ejemplo, usemos una estructura de red completamente conectada con tres capas.

Las capas completamente conectadas se definen mediante la clase Dense.

Puede especificar el número de neuronas o nodos en la capa como primer argumento y la función de activación utilizando el argumento de activación.

**Además, utilizaremos la función de activación de unidad lineal rectificada denominada ReLU en las dos primeras capas y la función Sigmoidea en la capa de salida.**

Solía darse el caso de que las funciones de activación Sigmoide y Tanh eran preferidas para todas las capas. Hoy en día, se logra un mejor rendimiento utilizando la función de activación ReLU.

**El uso de un sigmoide en la capa de salida garantiza que la salida de su red esté entre 0 y 1 y es fácil de asignar a una probabilidad de clase 1 o ajustarse a una clasificación estricta de cualquier clase con un umbral predeterminado de 0.5.**

En resumen:

* El modelo espera filas de datos con 5 variables (el argumento input_shape =(5,)).
* La primera capa oculta tiene 12 nodos y utiliza la función de activación ReLU.
* La segunda capa oculta tiene 5 nodos y utiliza la función de activación ReLU.
* La capa de salida tiene un nodo y utiliza la función de activación sigmoide.

In [13]:
#
model = Sequential()

model.add(Dense(12, input_shape = (6,), activation = 'relu'))

model.add(Dense(6, activation = 'relu'))

model.add(Dense(1, activation = 'sigmoid'))

## Compile Keras Model

La compilación del modelo utiliza bibliotecas numéricas eficientes (el llamado backend), como Theano o TensorFlow.

El backend elige automáticamente la mejor manera de representar la red para entrenar y hacer predicciones para ejecutar en su hardware, como CPU, GPU o incluso distribuido.

Al compilar, se debe especificar algunas propiedades adicionales necesarias al entrenar la red. Recuerde entrenar una red significa encontrar el mejor conjunto de ponderaciones para asignar entradas a salidas en su conjunto de datos.

Debe especificar la función de pérdida que se utilizará para evaluar un conjunto de ponderaciones, el optimizador utilizado para buscar diferentes ponderaciones para la red y cualquier métrica opcional que desee recopilar e informar durante el entrenamiento.

En este caso, utilizamos la **entropía cruzada** como argumento de pérdida. Esta pérdida es para problemas de clasificación binaria y se define en Keras como "binary_crossentropy".

Definiremos el optimizador como el eficiente algoritmo de descenso de gradiente estocástico "adam". Esta es una versión popular del descenso de gradiente porque se sintoniza automáticamente y da buenos resultados en una amplia gama de problemas.

Finalmente, debido a que se trata de un problema de clasificación, recopilará e informará la precisión de la clasificación definida mediante el argumento de métricas.

In [14]:
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

## Fit Keras Model

Ahora es el momento de ejecutar el modelo con algunos datos.

Entrenamos o ajustamos el modelo a los datos cargados llamando a la función fit() en el modelo.

El entrenamiento se produce a lo largo de épocas (epoch) y cada época se divide en lotes (batches).

* Época: una pasada por todas las filas del conjunto de datos de entrenamiento

* Lote: una o más muestras consideradas por el modelo dentro de una época antes de que se actualicen los pesos.

**Una época comprende uno o más lotes, según el tamaño de lote elegido, y el modelo es apto para muchas épocas.**

El proceso de entrenamiento se ejecutará durante un número fijo de épocas (iteraciones) a través del conjunto de datos que debe especificar mediante el argumento epochs.

Se debe establecer el número de filas del conjunto de datos que se consideran antes de que se actualicen los pesos del modelo dentro de cada época, denominado tamaño de lote, y establecerlo mediante el argumento batch_size.

Nuestro problema se ejecutará durante una pequeña cantidad de épocas (150) y utilizará un tamaño de lote relativamente pequeño de 10.

**Estas configuraciones se pueden elegir experimentalmente mediante prueba y error.**

In [15]:
# fit the keras model on the dataset
model.fit(X, y, epochs=150, batch_size=10)

Epoch 1/150
630/630 [==============================] - 2s 2ms/step - loss: 53.5233 - accuracy: 0.7467
Epoch 2/150
630/630 [==============================] - 1s 2ms/step - loss: 0.5309 - accuracy: 0.8335
Epoch 3/150
630/630 [==============================] - 1s 2ms/step - loss: 0.6234 - accuracy: 0.8230
Epoch 4/150
630/630 [==============================] - 1s 2ms/step - loss: 0.5744 - accuracy: 0.8273
Epoch 5/150
630/630 [==============================] - 1s 2ms/step - loss: 0.6486 - accuracy: 0.8210
Epoch 6/150
630/630 [==============================] - 2s 3ms/step - loss: 0.5674 - accuracy: 0.8306
Epoch 7/150
630/630 [==============================] - 2s 3ms/step - loss: 0.5713 - accuracy: 0.8262
Epoch 8/150
630/630 [==============================] - 1s 2ms/step - loss: 0.6792 - accuracy: 0.8246
Epoch 9/150
630/630 [==============================] - 1s 2ms/step - loss: 0.5315 - accuracy: 0.8294
Epoch 10/150
630/630 [==============================] - 1s 2ms/step - loss: 0.6077 - accur

## Evaluate Keras Model

Una vez entrenada nuestra red neuronal en todo el conjunto de datos y podemos evaluar el rendimiento de la red en el mismo conjunto de datos.

Esto nos dará una idea de qué tan bien ha modelado el conjunto de datos, pero no da idea de qué tan bien podría funcionar el algoritmo con datos nuevos.

**Esto se hizo por simplicidad, pero lo ideal sería separar los datos en conjuntos de datos de entrenamiento y prueba para entrenar y evaluar su modelo.**

Podemos evaluar el modelo en el conjunto de datos de entrenamiento usando la función evaluate() y pasarle la misma entrada y salida utilizada para entrenar el modelo.

Esto generará una predicción para cada par de entrada y salida y recopilará puntuaciones, incluida la pérdida promedio y cualquier métrica que haya configurado, como la precisión.

La función evaluate() devolverá una lista con dos valores. El primero será la pérdida del modelo en el conjunto de datos y el segundo será la precisión del modelo en el conjunto de datos. Habitualmente sólo interesa informar la precisión, así que podemos ignorar el valor de la pérdida.

In [16]:
# evaluate the keras model
_, accuracy = model.evaluate(X, y)
print('Accuracy: %.2f' % (accuracy*100))

197/197 [==============================] - 0s 2ms/step - loss: 0.2643 - accuracy: 0.8759
Accuracy: 87.59


Este resultado obtenido indica que el modelo de Keras ha sido evaluado en un conjunto de datos de **197 muestras**, y ha obtenido una **pérdida de 0.2643** y una **precisión del 87.59%**.

Esto sugiere un buen rendimiento del modelo en la tarea de clasificación o predicción para la que fue entrenado.

La precisión del **87.59%** es un valor *relativamente alto*, lo que significa que el modelo está *clasificando o prediciendo correctamente en aproximadamente el 87.59% de los* casos en el conjunto de datos de evaluación de la precepción de mujeres violentadas en países sudamericanos.

## Nota importante:

Si intenta ejecutar este ejemplo en una computadora portátil IPython o Jupyter, es posible que obtenga un error.

El motivo son las barras de progreso de salida durante el entrenamiento. Puede desactivarlos fácilmente configurando verbose=0 en la llamada a las funciones fit() y evalua(); Por ejemplo:

In [17]:
# fit the keras model on the dataset without progress bars
model.fit(X, y, epochs=150, batch_size=10, verbose=0)
# evaluate the keras model
_, accuracy = model.evaluate(X, y, verbose=0)

## Make Predictions

In [18]:
# make probability predictions with the model
predictions = model.predict(X)

predictions[:10]

197/197 [==============================] - 0s 1ms/step


array([[2.2554575e-03],
       [6.0567832e-05],
       [5.4873366e-05],
       [3.3977933e-05],
       [1.4100157e-05],
       [1.3939024e-05],
       [7.3739070e-06],
       [3.1000754e-06],
       [2.6231492e-06],
       [2.4039991e-06]], dtype=float32)

In [19]:
# round predictions
rounded = [round(x[0]) for x in predictions]

rounded[:10]

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

In [20]:
# Alternately, you can convert the probability into 0 or 1 to predict crisp classes directly; for example:

# make class predictions with the model
predictions = (model.predict(X) > 0.5).astype(int)

predictions[:10]

197/197 [==============================] - 0s 2ms/step


array([[0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0]])